src/postprocess/stresses.jl

```diff
@@ -26,12 +25,38 @@ for details. In Voigt notation one would use the vector
                                    basis.use_symmetries_for_kpoint_reduction,
                                    basis.comm_kpts, basis.architecture)
         ρ = compute_density(new_basis, scfres.ψ, scfres.occupation)
-        energies = energy_hamiltonian(new_basis, scfres.ψ, scfres.occupation;
-                                      ρ, scfres.eigenvalues, scfres.εF).energies
+        (; energies) = energy(new_basis, scfres.ψ, scfres.occupation;
+                              ρ, scfres.eigenvalues, scfres.εF)
         energies.total
     end
-    L = scfres.basis.model.lattice
-    Ω = scfres.basis.model.unit_cell_volume
-    stresses = ForwardDiff.gradient(M -> HF_energy((I+M) * L), zero(L)) / Ω
-    symmetrize_stresses(scfres.basis, stresses)
+    L  = scfres.basis.model.lattice
+    Ω  = scfres.basis.model.unit_cell_volume
+
+    # Define f(ϵ) = E[ (I+ϵ) * L]. Since the strain is symmetric (same as σ) it has only
+    # 6 free components which we collect as in Voigt notation
+    #    M = [ϵ_{xx} ϵ_{yy} ϵ_{zz} ϵ_{zy} ϵ_{zx} ϵ_{yx}]
+    # Then
+    function HF_energy_voigt(M)
+        D = [1+M[1]   M[6]   M[5];  # Lattice distortion matrix
+               M[6] 1+M[2]   M[4];
+               M[5]   M[4] 1+M[3]]
+        HF_energy(D * L)
+    end
+    # The derivative of this function wrt. M is by the chain rule
+    #    [df/dϵ_{xx}, df/dϵ_{yy}, df/dϵ_{zz},
+    #     df/dϵ_{zy}+df/dϵ_{yz}, df/dϵ_{zx}+df/dϵ_{xz}, df/dϵ_{yx}+df/dϵ_{xy}]
+    # Therefore
+    stress_voigt = 1/Ω * ForwardDiff.gradient(HF_energy_voigt, zeros(eltype(L), 6))
+    symmetrize_stresses(scfres.basis, voigt_to_full(stress_voigt))
+end
+function voigt_to_full(v::AbstractVector{T}) where {T}
```

Comment:
Still not clear to me the relationship between voigt_to_full and HF_energy_voigt, and why HF_energy_voigt doesn't just call voigt_to_full. Feels like what's in HF_energy_voigt is the "true" voigt_to_full function and the current voigt_to_full function is something else: the adjoint of full_to_voigt (full_to_voigt being the inverse of voigt_to_full).

Reply:
So I think just rename? Have voigt_to_full be what's in HF_energy_voigt, have full_to_voigt be the inverse of that, and define adjoint_full_to_voigt be what's in voigt_to_full now.

Reply:
Oh gosh. I had a bit of a muddle in my head. Turns out strain and stress Voigt notation are actually not the same and this is where my confusion came from. I now added separate transformation routines for both strain and stress. Should be clearer now.

Reply:
Yes the relationship should be that one is the adjoint of the inverse of the other, good you clarified